spark3/src/main/java/org/apache/iceberg/spark/FileRewriteCoordinator.java

```diff
@@ -0,0 +1,155 @@
+    LOG.info("Staging results of rewriting file set {} for table {}", fileSetID, table);
+    Pair<String, String> id = toID(table, fileSetID);
+    resultMap.put(id, newDataFiles);
+  }
+
+  public void commitRewrite(Table table, String fileSetID) {
+    commitRewrite(table, ImmutableSet.of(fileSetID));
+  }
+
+  public void commitRewrite(Table table, Set<String> fileSetIDs) {
+    LOG.info("Committing rewrite of file sets {} for table {}", fileSetIDs, table);
+
+    Set<DataFile> rewrittenDataFiles = fetchRewrittenDataFiles(table, fileSetIDs);
+    Set<DataFile> newDataFiles = fetchNewDataFiles(table, fileSetIDs);
+
+    table.newRewrite()
+        .rewriteFiles(rewrittenDataFiles, newDataFiles)
+        .commit();
+
+    LOG.info("Successfully committed rewrite of file sets {} for table {}", fileSetIDs, table);
+  }
+
+  private Set<DataFile> fetchRewrittenDataFiles(Table table, Set<String> fileSetIDs) {
+    FileScanTaskSetManager taskSetManager = FileScanTaskSetManager.get();
+
+    Set<DataFile> rewrittenDataFiles = Sets.newHashSet();
+
+    for (String fileSetID : fileSetIDs) {
+      List<FileScanTask> tasks = taskSetManager.fetchTasks(table, fileSetID);
+      ValidationException.check(tasks != null,
+          "Task set manager has no tasks for table %s with id %s",
+          table, fileSetID);
+
+      for (FileScanTask task : tasks) {
+        DataFile dataFile = task.file();
+        rewrittenDataFiles.add(dataFile);
+      }
+    }
+
+    return Collections.unmodifiableSet(rewrittenDataFiles);
+  }
+
+  private Set<DataFile> fetchNewDataFiles(Table table, Set<String> fileSetIDs) {
+    List<Set<DataFile>> results = Lists.newArrayList();
+
+    for (String fileSetID : fileSetIDs) {
+      Pair<String, String> id = toID(table, fileSetID);
+      Set<DataFile> result = resultMap.get(id);
+      ValidationException.check(result != null,
+          "No results for rewrite of file set %s in table %s",
+          fileSetID, table);
+
+      results.add(result);
+    }
+
+    Set<DataFile> newDataFiles = results.get(0);
+    for (int index = 1; index < results.size(); index++) {
+      newDataFiles = Sets.union(newDataFiles, results.get(index));
+    }
+
+    return newDataFiles;
+  }
+
+  public void abortRewrite(Table table, String fileSetID) {
+    LOG.info("Aborting rewrite of file set {} for table {}", fileSetID, table);
+    Pair<String, String> id = toID(table, fileSetID);
+    Set<DataFile> dataFiles = resultMap.remove(id);
+    if (dataFiles != null) {
+      LOG.info("Deleting {} uncommitted data files", dataFiles.size());
+      deleteFiles(table.io(), dataFiles);
+    }
+  }
+
+  public void abortRewrite(Table table, Set<String> fileSetIDs) {
+    for (String fileSetID : fileSetIDs) {
+      abortRewrite(table, fileSetID);
+    }
+  }
+
+  private void deleteFiles(FileIO io, Iterable<DataFile> dataFiles) {
+    Tasks.foreach(dataFiles)
```

Comment:
Do we usually do this single threaded?

Reply:
I think the current codebase is not consistent. We can refine this later if needed by passing an executor service here.